### ISSUE TITLE
twisted.logger.Logger.failure writes non-JSON to JSON(-seq) logs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8858]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)| @Julian reported|
|-|-|
|Trac ID|trac#8858|
|Type|defect|
|Created|2016-10-11 21:06:31Z|

Given that failure is somewhat special cased:

https://github.com/twisted/twisted/blob/e38cc25a67747899c6984d6ebaa8d3d134799415/src/twisted/logger/_file.py#L55-L60

the observer doesn't get a chance to look at it. So if you have an e.g. JSON output log, it will have tracebacks interspersed in it (breaking e.g. jsonEventsFromLogFile).

Sample:
```
import sys
from twisted.logger import Logger, globalLogPublisher, jsonFileLogObserver
print globalLogPublisher._observers[0]._observers
globalLogPublisher.addObserver(jsonFileLogObserver(open("output", "w")))
print globalLogPublisher._observers[0]._observers
l = Logger()
try:
    raise IndexError
except Exception:
    l.failure("asdf")
```

which will produce non-JSON in that file.

(Or really, it doesn't use the attached observer to format events, which I think probably will apply no matter what observer you used)

<details><summary>Searchable metadata</summary>

```
trac-id__8858 8858
type__defect defect
reporter__Julian Julian
priority__normal normal
milestone__New_Logging_System New Logging System
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__logger logger
keywords__None None
time__1476219991492905 1476219991492905
changetime__1481166108934498 1481166108934498
version__None None
owner__Glyph__glyph_____ Glyph <glyph@...>
cc__wsanchez@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_2"></a>|@wsanchez commented|
|-|-|

Agreed that what we're doing here is assuming "classic log text" in the `FileLogObserver` class, which is a fail…

One option would be to have `FileLogObserver` edit the format string to insert the traceback, but that's altering a key part of the event, and, in fact, the format string is one way to identify the type of event, so altering it would seem rude.

So perhaps that logic belongs in `formatEventAsClassicLogText`.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Replying to [wsanchez](#note_2):

> So perhaps that logic belongs in `formatEventAsClassicLogText`.

Agreed; that's where I started putting the fix even before reading this comment.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph commented|
|-|-|

Pull request available at https://github.com/twisted/twisted/pull/560.  Hoping the build succeeds :).

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_5"></a>|@markrwilliams commented|
|-|-|

https://github.com/twisted/twisted/pull/560#pullrequestreview-5358875

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph commented|
|-|-|

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @glyph<br>@glyph set status to `closed`|
|-|-|

In changeset 8b5865cda742b7d727f97cdec4b9e83974d57c94
```
#!CommitTicketReference repository="" revision="8b5865cda742b7d727f97cdec4b9e83974d57c94"
Merge 8858-tracebacks: fix JSON traceback format

Author: glyph

Reviewer: runciter

Fixes: #8858

twisted.logger.jsonFileLogObserver no longer emits non-JSON tracebacks into its file; additionally, twisted.logger.formatEventAsClassicLogText now includes traceback information for the log event it formats.
```
